bindings/sundials4py/examples/arkode/ark_brusselator.py

```diff
@@ -0,0 +1,227 @@
+        ydot[1] = w * u - v * u * u
+        ydot[2] = (b - w) / ep - w * u
+        return 0
+
+    def jac(self, t, yvec, fyvec, J, tmp1, tmp2, tmp3):
+        y = N_VGetArrayPointer(yvec)
+        a, b, ep = self.a, self.b, self.ep
+        u, v, w = y[0], y[1], y[2]
+        Jdata = SUNDenseMatrix_Data(J)
+        Jdata[0, 0] = -(w + 1.0) + 2.0 * u * v
+        Jdata[0, 1] = u * u
+        Jdata[0, 2] = -u
+        Jdata[1, 0] = w - 2.0 * u * v
+        Jdata[1, 1] = -u * u
+        Jdata[1, 2] = u
+        Jdata[2, 0] = -w
+        Jdata[2, 1] = 0.0
+        Jdata[2, 2] = -1.0 / ep - u
+        return 0
+
+
+def main():
+    # Problem parameters for Test 2
+    u0 = 1.2
+    v0 = 3.1
+    w0 = 3.0
+    a = 1.0
+    b = 3.5
+    ep = 5.0e-6
+    T0 = 0.0
+    Tf = 10.0
+    dTout = 1.0
+    NEQ = 3
+    Nt = int(np.ceil(Tf / dTout))
+    reltol = 1.0e-6
+    abstol = 1.0e-10
+
+    status, sunctx = SUNContext_Create(SUN_COMM_NULL)
+    y = N_VNew_Serial(NEQ, sunctx)
+
+    # Create ODE problem instance and set initial conditions
+    ode_problem = BrusselatorODE(u0, v0, w0, a, b, ep)
+    ode_problem.set_init_cond(y)
+
+    ark = ARKStepCreate(
+        None,  # f_E (explicit)
+        lambda t, yvec, ydotvec, _: ode_problem.f(t, yvec, ydotvec),  # f_I (implicit)
+        T0,
+        y,
+        sunctx,
+    )
+
+    status = ARKodeSStolerances(ark.get(), reltol, abstol)
+    assert status == ARK_SUCCESS
+
+    status = ARKodeSetInterpolantType(ark.get(), ARK_INTERP_LAGRANGE)
+    assert status == ARK_SUCCESS
+
+    status = ARKodeSetDeduceImplicitRhs(ark.get(), 1)
+    assert status == ARK_SUCCESS
+
+    # Dense matrix and linear solver
+    A = SUNDenseMatrix(NEQ, NEQ, sunctx)
+    LS = SUNLinSol_Dense(y, A, sunctx)
+
+    status = ARKodeSetLinearSolver(ark.get(), LS, A)
+    assert status == ARK_SUCCESS
+
+    status = ARKodeSetJacFn(
+        ark.get(),
+        lambda t, yvec, fyvec, J, tmp1, tmp2, tmp3, _: ode_problem.jac(
+            t, yvec, fyvec, J, tmp1, tmp2, tmp3
+        ),
+    )
+    assert status == ARK_SUCCESS
+
+    # Signal that this problem does not explicitly depend on time
+    status = ARKodeSetAutonomous(ark.get(), 1)
+    assert status == ARK_SUCCESS
+
+    # Initial problem output
+    yarr = N_VGetArrayPointer(y)
+    print("\nBrusselator ODE test problem:")
+    print(f"    initial conditions:  u0 = {u0},  v0 = {v0},  w0 = {w0}")
+    print(f"    problem parameters:  a = {a},  b = {b},  ep = {ep}")
+    print(f"    reltol = {reltol},  abstol = {abstol}\n")
+    print("        t           u           v           w")
+    print("   -------------------------------------------")
+    print(f"  {T0:10.6f}  {yarr[0]:10.6f}  {yarr[1]:10.6f}  {yarr[2]:10.6f}")
+
+    # Main time-stepping loop: calls ARKodeEvolve to perform the integration,
+    # then prints results.  Stops when the final time has been reached. The
+    # solution is written out to a file.
+    with open("solution.txt", "w") as UFID:
```

Comment:
The reason the C example writes to file is so that the solution can be plotted (using Python).  It would be a more exciting example if the solutions were stored directly in a 2D numpy array and plotted at the end.

Reply:
It would be, but the idea was to do exactly what the C example did.

Reply:
I agree it would be good to come back after this PR and update some things (e.g., IO) to take advantage of python more and be too tied to what the C version does.